### ISSUE TITLE
Refactor SpiderLoader class for more flexible subclassing

### ISSUE THREAD

**Issue by lagenar** [scrapy/scrapy#1805]
Hi there,

This is not really an issue but just a suggestion to consider to make it easier to subclass the SpiderLoader class. I have a very big list of spiders and what I wanted to do is to optimize how spiders are searched on the file system by caching the location, and also because I want to avoid all spider modules to get imported.

What I found on the __init__ method of that class is:
```
    def __init__(self, settings):
            self.spider_modules = settings.getlist('SPIDER_MODULES')
            self._spiders = {}
            for name in self.spider_modules:
                for module in walk_modules(name):
                    self._load_spiders(module)`
```
The problem here is that the constructor is doing too much work and it's a bit difficult to override the default behaviour. Anyway, I think I will just implement my own class without sub classing but looks like this could be refactored a bit.

**Comment by kmike** [comment]
Do you have suggestions about how to improve the code?

I think that's fine to extract spider loading code to a method (`_load_all_spiders`? other ideas?), so that you don't have to copy-paste first two lines when overriding `__init__`. I also don't quite like that `self._load_spiders` mutates state, but that's another question :)

**Comment by lagenar** [comment]
Hi @kmike, yes looks like implementing the _load_all_spiders method is the most straightforward solution :) that would allow to subclass without touching the constructor.

**Comment by redapple** [comment]
Fixed in #1818
